# HW2 - Data Ingestion and Analysis

# Data Ingestion

#### Import Packages

In [2]:
import pandas as pd
from tqdm import tqdm
import json, time, urllib.parse
import requests

#### Get US cities and related articles data

In [2]:
cities_wiki_df = pd.read_csv('us_cities_by_state_SEPT.2023.csv')

state           page_title  \
0      Alabama   Abbeville, Alabama   
1      Alabama  Adamsville, Alabama   
2      Alabama     Addison, Alabama   
3      Alabama       Akron, Alabama   
4      Alabama   Alabaster, Alabama   
...        ...                  ...   
22152  Wyoming   Wamsutter, Wyoming   
22153  Wyoming   Wheatland, Wyoming   
22154  Wyoming     Worland, Wyoming   
22155  Wyoming      Wright, Wyoming   
22156  Wyoming       Yoder, Wyoming   

                                                     url  
0       https://en.wikipedia.org/wiki/Abbeville,_Alabama  
1      https://en.wikipedia.org/wiki/Adamsville,_Alabama  
2         https://en.wikipedia.org/wiki/Addison,_Alabama  
3           https://en.wikipedia.org/wiki/Akron,_Alabama  
4       https://en.wikipedia.org/wiki/Alabaster,_Alabama  
...                                                  ...  
22152   https://en.wikipedia.org/wiki/Wamsutter,_Wyoming  
22153   https://en.wikipedia.org/wiki/Wheatland,_Wyoming  
22154     https://en.wikipedia.org/wiki/Worland,_Wyoming  
22155      https://en.wikipedia.org/wiki/Wright,_Wyoming  
22156       https://en.wikipedia.org/wiki/Yoder,_Wyoming  

[22157 rows x 3 columns]

### Page Info Extraction using API

In [4]:
#########
#
#    CONSTANTS
#

# The basic English Wikipedia API endpoint
API_ENWIKIPEDIA_ENDPOINT = "https://en.wikipedia.org/w/api.php"

# We'll assume that there needs to be some throttling for these requests - we should always be nice to a free data resource
API_LATENCY_ASSUMED = 0.002       # Assuming roughly 2ms latency on the API and network
API_THROTTLE_WAIT = (1.0/100.0)-API_LATENCY_ASSUMED

# When making automated requests we should include something that is unique to the person making the request
# This should include an email - your UW email would be good to put in there
REQUEST_HEADERS = {
    'User-Agent': '<uwnetid@uw.edu>, University of Washington, MSDS DATA 512 - AUTUMN 2023',
}

# This is just a list of English Wikipedia article titles that we can use for example requests
ARTICLE_TITLES = [ 'Bison', 'Northern flicker', 'Red squirrel', 'Chinook salmon', 'Horseshoe bat' ]

# This is a string of additional page properties that can be returned see the Info documentation for
# what can be included. If you don't want any this can simply be the empty string
PAGEINFO_EXTENDED_PROPERTIES = "talkid|url|watched|watchers"
#PAGEINFO_EXTENDED_PROPERTIES = ""

# This template lists the basic parameters for making this
PAGEINFO_PARAMS_TEMPLATE = {
    "action": "query",
    "format": "json",
    "titles": "",           # to simplify this should be a single page title at a time
    "prop": "info",
    "inprop": PAGEINFO_EXTENDED_PROPERTIES
}


The API request will be made using one procedure. The idea is to make this reusable. The procedure is parameterized, but relies on the constants above for the important parameters. The underlying assumption is that this will be used to request data for a set of article pages. Therefore the parameter most likely to change is the article_title.

In [5]:
#########
#
#    PROCEDURES/FUNCTIONS
#

def request_pageinfo_per_article(article_title = None, 
                                 endpoint_url = API_ENWIKIPEDIA_ENDPOINT, 
                                 request_template = PAGEINFO_PARAMS_TEMPLATE,
                                 headers = REQUEST_HEADERS):
    
    # article title can be as a parameter to the call or in the request_template
    if article_title:
        request_template['titles'] = article_title

    if not request_template['titles']:
        raise Exception("Must supply an article title to make a pageinfo request.")

    # make the request
    try:
        # we'll wait first, to make sure we don't exceed the limit in the situation where an exception
        # occurs during the request processing - throttling is always a good practice with a free
        # data source like Wikipedia - or any other community sources
        if API_THROTTLE_WAIT > 0.0:
            time.sleep(API_THROTTLE_WAIT)
        response = requests.get(endpoint_url, headers=headers, params=request_template)
        json_response = response.json()
    except Exception as e:
        print(e)
        json_response = None
    return json_response


In [9]:
article_titles = list(cities_wiki_df['page_title'])

Creating a function to get page info for each article title. We will pass article title to this function and it will store the information related to the article including the revision ID into a dictionary for all 22,157 articles.

In [10]:
# Function Definition
def get_page_info(df):
    article_title = df['page_title']
    try:
        info = request_pageinfo_per_article(article_title)
        page_dict = info['query']['pages']
        all_page_info_dict.update(page_dict)
    except:
        print(article_title)

In [11]:
# Function Call
all_page_info_dict = dict()
cities_wiki_df.apply(get_page_info,axis=1)

0        None
1        None
2        None
3        None
4        None
         ... 
22152    None
22153    None
22154    None
22155    None
22156    None
Length: 22157, dtype: object

In [16]:
page_info_df = pd.DataFrame(all_page_info_dict).T.reset_index(drop=True)

In [17]:
page_info_df.to_csv('page_info_df.csv')

In [3]:
page_info_df = pd.read_csv('page_info_df.csv')

In [4]:
page_info_df_final = page_info_df[['pageid','title','lastrevid']]

pageid                title   lastrevid
0      104730   Abbeville, Alabama  1171163550
1      104761  Adamsville, Alabama  1177621427
2      105188     Addison, Alabama  1168359898
3      104726       Akron, Alabama  1165909508
4      105109   Alabaster, Alabama  1179139816
...       ...                  ...         ...
21514  140221   Wamsutter, Wyoming  1169591845
21515  140185   Wheatland, Wyoming  1176370621
21516  140245     Worland, Wyoming  1166347917
21517  140070      Wright, Wyoming  1166334449
21518  140112       Yoder, Wyoming  1171182284

[21519 rows x 3 columns]

## ORES Score extraction using API

In [6]:
#########
#
#    CONSTANTS
#

#    The current LiftWing ORES API endpoint and prediction model
#
API_ORES_LIFTWING_ENDPOINT = "https://api.wikimedia.org/service/lw/inference/v1/models/{model_name}:predict"
API_ORES_EN_QUALITY_MODEL = "enwiki-articlequality"

#
#    The throttling rate is a function of the Access token that you are granted when you request the token. The constants
#    come from dissecting the token and getting the rate limits from the granted token. An example of that is below.
#
API_LATENCY_ASSUMED = 0.002       # Assuming roughly 2ms latency on the API and network
API_THROTTLE_WAIT = (60.0/5000.0)-API_LATENCY_ASSUMED

#    When making automated requests we should include something that is unique to the person making the request
#    This should include an email - your UW email would be good to put in there
#    
#    Because all LiftWing API requests require some form of authentication, you need to provide your access token
#    as part of the header too
#
REQUEST_HEADER_TEMPLATE = {
    'User-Agent': "<nshah23@uw.edu>, University of Washington, MSDS DATA 512 - AUTUMN 2023",
    'Content-Type': 'application/json',
    'Authorization': "Bearer {access_token}"
}
#
#    This is a template for the parameters that we need to supply in the headers of an API request
#
REQUEST_HEADER_PARAMS_TEMPLATE = {
    'email_address' : "",         # your email address should go here
    'access_token'  : ""          # the access token you create will need to go here
}

#
#    A dictionary of English Wikipedia article titles (keys) and sample revision IDs that can be used for this ORES scoring example
#
ARTICLE_REVISIONS = { 'Bison':1085687913 , 'Northern flicker':1086582504 , 'Red squirrel':1083787665 , 'Chinook salmon':1085406228 , 'Horseshoe bat':1060601936 }

#
#    This is a template of the data required as a payload when making a scoring request of the ORES model
#
ORES_REQUEST_DATA_TEMPLATE = {
    "lang":        "en",     # required that its english - we're scoring English Wikipedia revisions
    "rev_id":      "",       # this request requires a revision id
    "features":    True
}

#
#    These are used later - defined here so they, at least, have empty values
#
USERNAME = ""
ACCESS_TOKEN = ""
#

In [7]:
USERNAME = "Nshah23"
ACCESS_TOKEN = "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJhdWQiOiIwMjExZWE0YzgxYjI1OTBlYjAyMzE4MzkwOGQ5NDRhYSIsImp0aSI6IjA0YjIxZGZiMjkzZWVjYmZkNmUxNmY2OWZkMTQ5YjYzODRiMzdmNTRiYzJmNTNlMWQxN2IwYzFhZDkzYzQwNjE0OGJiNTc4MmZiM2VlNmQ2IiwiaWF0IjoxNjk3NDk5MzY4LjI0NzM0NywibmJmIjoxNjk3NDk5MzY4LjI0NzM1LCJleHAiOjMzMjU0NDA4MTY4LjI0NDc3NCwic3ViIjoiNzQwMjE2MTIiLCJpc3MiOiJodHRwczovL21ldGEud2lraW1lZGlhLm9yZyIsInJhdGVsaW1pdCI6eyJyZXF1ZXN0c19wZXJfdW5pdCI6NTAwMCwidW5pdCI6IkhPVVIifSwic2NvcGVzIjpbImJhc2ljIl19.CB6-OrWFJZ7DMAyv_Iktz7UKL_e2Wz_ZeFY6lZQUREpgntObptjYq-GKgKVfKXfEE19HZTg_RCCe4rbQ5mFDcCP9Jg5TFXQWOE8ot5QCi4dnOPjlV0DCWNmJQQRU7ulwvl1Bqzmt33REU9n9FszvZ65vaRKfHq_leB7SB-Yldr9qUfsRqJ3nUDLPXer6NaFwR_YUPP-9gh225SgMgjW-_6n-vDYBtrw-3WL4PcXbZImcMA-J0_0QNUA-LyfwqodmCSIXexOMFFUkv7mj7Rz-q-Qi-g9pehAwvVV3WKY3bxWX99KWGqVINnP8UBP7K0lMX9qr5QtJG-n6r28hRTEU6GjjzKYxfGhW6-g1_LibWy-J4QXSqAW8sRooYrhTrTfQeKa9EucgHO3390G4xI4et_KIJT1rChvmPh5_c_eK69EJeQ36KDMsoO2le9SjRh0ed_9hd55f3xzix3G0H-8sZMFkhbLlGATuJvRVpWwbOSUhBTMD8CBTfCNegFSY894bR9rSFyEYEf9EeYCPQh5ICccrX4vxOKiloBvF1wcXcdOyhYe62MBRKgGDSs01tg9dY1gMJy6rEIAvVb-qEW1mvkUoUPFSh-bUoIkOA9rRzSmpbyAWcyEZbCr4aKJHQftDDdTHNHJFNFbO6ipWsk8Jt_PRIjwG_sMun-3cWFTx4V0"

In [9]:
#
#   Decode the Wikimedia JWT Access token
#
#   NOTE: This is not required to use LiftWing to request ORES scores. This is just being done to satisfy my curiosity.
#
import base64

print("Decoding the ACCESS_TOKEN:")
try:
    token_components = ACCESS_TOKEN.split(".")
    if len(token_components) == 3:
        header = json.loads(base64.b64decode(token_components[0]).decode())
        payload = json.loads(base64.b64decode(token_components[1]).decode())
        print("Token Header:",json.dumps(header,indent=4))
        print("Token Payload:",json.dumps(payload,indent=4))
        #print("Token Signature:",token_components[2])
        print("Token Signature: <value_suppressed>")
        #
        #  One should be able to use public/private keys to actually validate that signature - left as an exercise for later
        #
    else:
        print(f"The ACCESS_TOKEN appears to be improperly structured. It should have 3 components and it has {len(token_components)}")
except Exception as ex:
    print(f"Looks like the ACCESS_TOKEN is undefined or an empty value")
    raise(ex)


Decoding the ACCESS_TOKEN:
Token Header: {
    "typ": "JWT",
    "alg": "RS256"
}
Token Payload: {
    "aud": "0211ea4c81b2590eb023183908d944aa",
    "jti": "04b21dfb293eecbfd6e16f69fd149b6384b37f54bc2f53e1d17b0c1ad93c406148bb5782fb3ee6d6",
    "iat": 1697499368.247347,
    "nbf": 1697499368.24735,
    "exp": 33254408168.244774,
    "sub": "74021612",
    "iss": "https://meta.wikimedia.org",
    "ratelimit": {
        "requests_per_unit": 5000,
        "unit": "HOUR"
    },
    "scopes": [
        "basic"
    ]
}
Token Signature: <value_suppressed>


In [10]:
#########
#
#    PROCEDURES/FUNCTIONS
#

def request_ores_score_per_article(article_revid = None, email_address=None, access_token=None,
                                   endpoint_url = API_ORES_LIFTWING_ENDPOINT, 
                                   model_name = API_ORES_EN_QUALITY_MODEL, 
                                   request_data = ORES_REQUEST_DATA_TEMPLATE, 
                                   header_format = REQUEST_HEADER_TEMPLATE, 
                                   header_params = REQUEST_HEADER_PARAMS_TEMPLATE):
    
    #    Make sure we have an article revision id, email and token
    #    This approach prioritizes the parameters passed in when making the call
    if article_revid:
        request_data['rev_id'] = article_revid
    if email_address:
        header_params['email_address'] = email_address
    if access_token:
        header_params['access_token'] = access_token
    
    #   Making a request requires a revision id - an email address - and the access token
    if not request_data['rev_id']:
        raise Exception("Must provide an article revision id (rev_id) to score articles")
    if not header_params['email_address']:
        raise Exception("Must provide an 'email_address' value")
    if not header_params['access_token']:
        raise Exception("Must provide an 'access_token' value")
    
    # Create the request URL with the specified model parameter - default is a article quality score request
    request_url = endpoint_url.format(model_name=model_name)
    
    # Create a compliant request header from the template and the supplied parameters
    headers = dict()
    for key in header_format.keys():
        headers[str(key)] = header_format[key].format(**header_params)
    
    # make the request
    try:
        # we'll wait first, to make sure we don't exceed the limit in the situation where an exception
        # occurs during the request processing - throttling is always a good practice with a free data
        # source like ORES - or other community sources
        if API_THROTTLE_WAIT > 0.0:
            time.sleep(API_THROTTLE_WAIT)
        #response = requests.get(request_url, headers=headers)
        response = requests.post(request_url, headers=headers, data=json.dumps(request_data))
        json_response = response.json()
    except Exception as e:
        print(e)
        json_response = None
    return json_response


Defining a Function to get the Quality Score (Prediction) for each article based on its current Revision ID. ORES API is used to extract scores by passing each article in the function using the apply function and then getting the scores out as a dictionary for evry  Revision ID.

In [11]:
# Function Definition
def get_ores(df):
    try:
        score = request_ores_score_per_article(article_revid=df['lastrevid'],
                                               email_address="nshah23@uw.edu",
                                               access_token=ACCESS_TOKEN)
        # print(score)
        # score_dict = score['scores']
        ores_dict[df['lastrevid']]=score['enwiki']['scores'][str(df['lastrevid'])]['articlequality']['score']
    except:
        noscore_list.append(df['title'])

In [ ]:
# Function Call
ores_dict = dict()
noscore_list = []
tqdm.pandas()
page_info_df_final.progress_apply(get_ores, axis=1)

In [ ]:
ores_df = pd.DataFrame(ores_dict).T
ores_df

# DATA MERGE

Getting all the dataframes and merging them for the final analysis

In [ ]:
page_info_df = pd.read_csv('page_info_df.csv')
page_info_df = page_info_df[['title','lastrevid']]

ores_df = pd.read_csv('request_ores_score_per_article_output.csv')

wiki_ores_df = pd.merge(page_info_df,ores_df,on="lastrevid",how="inner")

us_cities_df = pd.read_csv('us_cities_by_state_SEPT.2023.csv')

wiki_final = pd.merge(us_cities_df,wiki_ores_df,left_on='page_title',right_on='title',how="left")

population_df = pd.read_csv('NST-EST2022-ALLDATA.csv')
population_df = population_df[['NAME','POPESTIMATE2022']]

all_merged_df = pd.merge(wiki_final,population_df,left_on="state",right_on="NAME",how="left")
all_merged_df['state'] = all_merged_df['state'].str.replace("_"," ")
all_merged_df['state'] = all_merged_df['state'].str.replace(" (U.S. state)","")

region_division_df = pd.read_excel('US States by Region - US Census Bureau.xlsx',index_col=[0,1])
region_division_df = region_division_df.droplevel(0).dropna().reset_index()
region_division_df.columns=['regional_division','state']

Merging all the datasets into a final dataframe.

In [ ]:
all_merged_df_division = pd.merge(all_merged_df,region_division_df,on="state",how="inner")
all_merged_df_division.drop(columns=['page_title','NAME','url'],inplace=True)
all_merged_df_division = all_merged_df_division.rename(columns={'prediction':'article_quality','POPESTIMATE2022':'population','title':'article_title','lastrevid':'revision_id'})
wp_scored_city_articles_by_state = all_merged_df_division[['state','regional_division','population','article_title','revision_id','article_quality']]

In [ ]:
wp_scored_city_articles_by_state.to_csv('wp_scored_city_articles_by_state.csv',index=False)

# ANALYSIS

In [ ]:
final_df = wp_scored_city_articles_by_state
state_df = final_df.groupby(['state','regional_division','population']).agg(total_articles=('article_title','count')).reset_index()
state_df['tot_articles_per_capita'] = state_df['total_articles']/state_df['population']
state_df

## Top 10 US states by coverage: The 10 US states with the highest total articles per capita (in descending order)

In [ ]:
top10states_by_coverage = state_df.sort_values('tot_articles_per_capita',ascending=False).head(10).reset_index(drop=True)
top10states_by_coverage

## Bottom 10 US states by coverage: The 10 US states with the lowest total articles per capita (in ascending order) .

In [ ]:
bottom10states_by_coverage = state_df.sort_values('tot_articles_per_capita').head(10).reset_index(drop=True)
bottom10states_by_coverage

In [ ]:
high_quality_list = ["FA","GA"]
high_quality_df = final_df[final_df['article_quality'].isin(high_quality_list)]
quality_df = high_quality_df.groupby(['state','regional_division','population']).agg(total_articles=('article_title','count')).reset_index()
quality_df['tot_articles_per_capita'] = quality_df['total_articles']/quality_df['population']
quality_df

## Top 10 US states by high quality: The 10 US states with the highest high quality articles per capita (in descending order) 

In [ ]:
top10states_by_quality = quality_df.sort_values('tot_articles_per_capita',ascending=False).head(10).reset_index(drop=True)
top10states_by_quality

## Bottom 10 US states by high quality: The 10 US states with the lowest high quality articles per capita (in ascending order).

In [ ]:
bottom10states_by_quality = quality_df.sort_values('tot_articles_per_capita').head(10).reset_index(drop=True)
bottom10states_by_quality

## Census divisions by total coverage: A rank ordered list of US census divisions (in descending order) by total articles per capita.

In [ ]:
division_df = final_df.groupby('regional_division').agg(total_articles=('article_title','count'),population=('population','sum')).reset_index()
division_df['tot_articles_per_capita'] = division_df['total_articles']/division_df['population']
census_divisions_by_total_coverage = division_df.sort_values('tot_articles_per_capita',ascending=False).reset_index(drop=True)
census_divisions_by_total_coverage

## Census divisions by high quality coverage: Rank ordered list of US census divisions (in descending order) by high quality articles per capita.

In [ ]:
high_quality_list = ["FA","GA"]
division_high_quality_df = final_df[final_df['article_quality'].isin(high_quality_list)]
division_quality_df = division_high_quality_df.groupby('regional_division').agg(total_articles=('article_title','count'),population=('population','sum')).reset_index()
division_quality_df['tot_articles_per_capita'] = division_quality_df['total_articles']/division_quality_df['population']
census_divisions_by_high_quality = division_quality_df.sort_values('tot_articles_per_capita',ascending=False).reset_index(drop=True)
census_divisions_by_high_quality